docs/source/extension/extension_dual_compatibility.rst

```diff
@@ -0,0 +1,107 @@
+.. Copyright (c) Jupyter Development Team.
+.. Distributed under the terms of the Modified BSD License.
+
+Building Lab 4 / Notebook 7 Compatible Jupyter Extensions
```

Comment:
Since this would apply to more lab-based applications than just JupyterLab and Notebook 7, maybe something like the following would work?

```suggestion
Building extensions for JupyterLab-based applications
```

Reply:
And then have a section focused on JupyterLab 4 and Notebook 7 since they are the two major notebook frontends at the moment built with this stack.

Reply:
This idea is going to be hard to communicate to people: NotebookFrontend is used to build Lab and NB7, other frontends may come out in the future, and that you can aim for compatibility with all of them. I think it would be easier to discuss Lab/NB7 first and then have a section later discussing other frontends that the other way around (kind of the reverse of what you're suggesting) plus a bit of minor rewording to accommodate these new concepts. Introducing all of these concepts up front I think is too heavy at the beginning of this document.

Reply:
> NotebookFrontend is used to build Lab and NB7

You mean `JupyterFrontEnd`?

Many extensions that provide commands, themes, menu entries, notebook toolbar buttons... don't need to care about the shell of the host application. Extensions should focus on Extension Points as defined in https://jupyterlab.readthedocs.io/en/latest/extension/extension_points.html.

Some applications might or might not provide some of these extension points, as it is the case with Notebook 7.

Trying to convey this idea in this document would be great so extension authors can follow best practices from the beginning. And would be coherent with the existing documentation.

Reply:
> You mean `JupyterFrontEnd`?

Yeah I meant this.

> Many extensions that provide commands, themes, menu entries, notebook toolbar buttons... don't need to care about the shell of the host application. Extensions should focus on Extension Points as defined in https://jupyterlab.readthedocs.io/en/latest/extension/extension_points.html.

I will definitely incorporate this, thanks for the feedback. I would probably call this the provider/consumer pattern here (I think this is how it is discussed in other parts of the docs already).

I still think it's useful to tell people how they can tailor specialized behaviors for a specific frontend by testing the kernel and using some thin frontend-specific UI code (and this is one of the approaches we discussed in the Notebook meeting). Though I can add some warnings/caveats about overusing this approach and how the other pattern should be the default.

> Some applications might or might not provide some of these extension points, as it is the case with Notebook 7.
> 
> Trying to convey this idea in this document would be great so extension authors can follow best practices from the beginning. And would be coherent with the existing documentation.

Will do!